{"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"B1PwSxMrOVag"},"outputs":[],"source":["import numpy as np\n","import matplotlib.pyplot as plt\n","import pandas as pd\n","import os\n","\n","from scipy.stats import zscore\n","\n","import networkx as nx\n"]},{"cell_type":"code","source":["#%% black list region and chip-seq peak files\n","blk_df = pd.read_csv('path/to/blk/blacklist.bed6format.bed', sep='\\t', header=None, )\n","blk_df\n"],"metadata":{"id":"hIhK8RkoOhj1","colab":{"base_uri":"https://localhost:8080/","height":423},"executionInfo":{"status":"ok","timestamp":1678496932463,"user_tz":480,"elapsed":1273,"user":{"displayName":"PY Wu","userId":"03546275994091638450"}},"outputId":"7f7238cf-df6e-485d-eea0-a6d30d1c9b12"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["        0         1         2                        3     4  5\n","0    chr1    564449    570371  High_Mappability_island  1000  .\n","1    chr1    724136    727043         Satellite_repeat  1000  .\n","2    chr1    825006    825115                 BSR/Beta  1000  .\n","3    chr1   2583334   2634374   Low_mappability_island  1000  .\n","4    chr1   4363064   4363242                 (CATTC)n  1000  .\n","..    ...       ...       ...                      ...   ... ..\n","406  chrY  28555026  28555353                     TAR1  1000  .\n","407  chrY  28784129  28819695         Satellite_repeat  1000  .\n","408  chrY  58819367  58917648                 (CATTC)n  1000  .\n","409  chrY  58971913  58997782                 (CATTC)n  1000  .\n","410  chrY  59361267  59362785                     TAR1  1000  .\n","\n","[411 rows x 6 columns]"],"text/html":["\n","  <div id=\"df-f369ea51-e75e-4d1e-91be-e8e0dafb32c6\">\n","    <div class=\"colab-df-container\">\n","      <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>0</th>\n","      <th>1</th>\n","      <th>2</th>\n","      <th>3</th>\n","      <th>4</th>\n","      <th>5</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>chr1</td>\n","      <td>564449</td>\n","      <td>570371</td>\n","      <td>High_Mappability_island</td>\n","      <td>1000</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>chr1</td>\n","      <td>724136</td>\n","      <td>727043</td>\n","      <td>Satellite_repeat</td>\n","      <td>1000</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>chr1</td>\n","      <td>825006</td>\n","      <td>825115</td>\n","      <td>BSR/Beta</td>\n","      <td>1000</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>chr1</td>\n","      <td>2583334</td>\n","      <td>2634374</td>\n","      <td>Low_mappability_island</td>\n","      <td>1000</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>chr1</td>\n","      <td>4363064</td>\n","      <td>4363242</td>\n","      <td>(CATTC)n</td>\n","      <td>1000</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>406</th>\n","      <td>chrY</td>\n","      <td>28555026</td>\n","      <td>28555353</td>\n","      <td>TAR1</td>\n","      <td>1000</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>407</th>\n","      <td>chrY</td>\n","      <td>28784129</td>\n","      <td>28819695</td>\n","      <td>Satellite_repeat</td>\n","      <td>1000</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>408</th>\n","      <td>chrY</td>\n","      <td>58819367</td>\n","      <td>58917648</td>\n","      <td>(CATTC)n</td>\n","      <td>1000</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>409</th>\n","      <td>chrY</td>\n","      <td>58971913</td>\n","      <td>58997782</td>\n","      <td>(CATTC)n</td>\n","      <td>1000</td>\n","      <td>.</td>\n","    </tr>\n","    <tr>\n","      <th>410</th>\n","      <td>chrY</td>\n","      <td>59361267</td>\n","      <td>59362785</td>\n","      <td>TAR1</td>\n","      <td>1000</td>\n","      <td>.</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>411 rows Ã— 6 columns</p>\n","</div>\n","      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-f369ea51-e75e-4d1e-91be-e8e0dafb32c6')\"\n","              title=\"Convert this dataframe to an interactive table.\"\n","              style=\"display:none;\">\n","        \n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","       width=\"24px\">\n","    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n","    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n","  </svg>\n","      </button>\n","      \n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      flex-wrap:wrap;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","      <script>\n","        const buttonEl =\n","          document.querySelector('#df-f369ea51-e75e-4d1e-91be-e8e0dafb32c6 button.colab-df-convert');\n","        buttonEl.style.display =\n","          google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","        async function convertToInteractive(key) {\n","          const element = document.querySelector('#df-f369ea51-e75e-4d1e-91be-e8e0dafb32c6');\n","          const dataTable =\n","            await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                     [key], {});\n","          if (!dataTable) return;\n","\n","          const docLinkHtml = 'Like what you see? Visit the ' +\n","            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","            + ' to learn more about interactive tables.';\n","          element.innerHTML = '';\n","          dataTable['output_type'] = 'display_data';\n","          await google.colab.output.renderOutput(dataTable, element);\n","          const docLink = document.createElement('div');\n","          docLink.innerHTML = docLinkHtml;\n","          element.appendChild(docLink);\n","        }\n","      </script>\n","    </div>\n","  </div>\n","  "]},"metadata":{},"execution_count":16}]},{"cell_type":"code","source":["#%% generate bed files for the eligible coordinates\n","# elegible coordinates should be (1) mappable, (2) not listed in black-list\n","# this part is for all chrs except chr9 and chr22\n","for chrid in ['chr{}'.format(i) for i in range(1,9)] + ['chr{}'.format(i) for i in range(10,22)] +['chrX']:\n","    print(chrid)\n","    hic_prob_df = pd.read_csv('path/to/save_folder/{}_K562_prob.5000.txt'.format(chrid),\n","                          sep='\\t', header=None)\n","\n","    # mappable : exist self interaction detection\n","    all_cord = list(hic_prob_df[hic_prob_df[4]==9][0]) # all coordinates that have self interactions, so these sites are mappable\n","\n","    # build blk list file, extract this chromosome, then round the coordinates to 5000\n","    blk_chr = blk_df[blk_df[0] == chrid].copy().iloc[:,1:3]\n","    blk_chr['s'] = blk_chr[1] // 5000 * 5000\n","    blk_chr['e'] = blk_chr[2] // 5000 * 5000\n","    blk_cords = []\n","    for s,e in blk_chr.iloc[:,2:].values:\n","        blk_cords.extend([c_ for c_ in range(s, e+5000, 5000)])\n","\n","    # blk list : remove intersect of `all_cord`  and `blk`\n","    eligible_coord = [c_ for c_ in all_cord if c_ not in blk_cords]\n","\n","    # convert list to bed format\n","    node_bed = pd.DataFrame({\n","        0: chrid,\n","        1: eligible_coord,\n","        2: np.array(eligible_coord)+5000\n","    })\n","\n","    print(node_bed.iloc[:5,])\n","\n","    # save bed files\n","    node_bed.to_csv(os.path.join('results/eligible_coordinates',\n","                                 'eligible_coordinates.{}.hg19.bed'.format(chrid)), sep='\\t', header=None, index=None, )\n","\n"],"metadata":{"id":"7Nfczd726mJB"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#%% generate bed files for the eligible coordinates\n","# elegible coordinates should be (1) mappable, (2) not listed in black-list\n","# this part is for chr9 and chr22\n","\n","\"\"\"\n","only chr9 and chr22 , all the other chr are the same as the ref genome\n","\n","ref:\n","    chr9: A----ABL1----B\n","    chr22: C----BCR----D\n","\n","\n","K562 transloc between chr9 and chr22 to generate:\n","    der9: chr9: A----BCR----D\n","    philchr22: chr22: C----ABL1----B\n","\n","\n","hg38:\n","\n","    t(9;22)(q34;q11)\n","    ABL1 on chr9: 130713016-130887670\n","    BCR on chr22: 23180509-23318037\n","\n","    ->\n","\n","    der9: chr9(0-130713016)-chr22(23318037-End)\n","    philchr22: chr22(0-23318037)-chr9(130713016-End)\n","    # all.hg38.fa only has chr1~22, X, Y, M curated\n","\n","\n","hg19:\n","    t(9;22)(q34;q11)\n","    ABL1 on chr9:133589268-133763062\n","    BCR on chr22:23522552-23660224\n","\n","    ->\n","    der9: chr9(0-133589268)-chr22(23660224-End)\n","    philchr22: chr22(0-23660224)-chr9(133589268-End)\n","\n","\"\"\"\n","\n","transloc_coord_dict = {\n","    'hg19': {\n","        'ABL1': [133589268,133763062],\n","        'BCR': [23522552,23660224],\n","    },\n","    'hg38': {\n","        'ABL1': [130713016,130887670],\n","        'BCR': [23180509,23318037],\n","    },\n","}\n","\n","transloc_gene_dict = {\n","    'chr9': 'ABL1',\n","    'chr22': 'BCR'\n","}\n","\n","res = 5000\n","assembly = 'hg19'\n","\n","for chrid in ['chr9', 'chr22']:\n","    print(chrid)\n","    transloc_gene = transloc_gene_dict[chrid]\n","    spl_s, _ = transloc_coord_dict[assembly][transloc_gene]\n","\n","    hic_prob_df = pd.read_csv('path/to/save_folder/{}_K562_prob.5000.txt'.format(chrid),\n","                          sep='\\t', header=None)\n","\n","    # mappable : exist self interaction detection\n","    all_cord = list(hic_prob_df[hic_prob_df[4]==9][0]) # all coordinates that have self interactions, so these sites are mappable\n","\n","    # build blk list file, extract this chromosome, then round the coordinates to 5000\n","    blk_chr = blk_df[blk_df[0] == chrid].copy().iloc[:,1:3]\n","    blk_chr['s'] = blk_chr[1] // 5000 * 5000\n","    blk_chr['e'] = blk_chr[2] // 5000 * 5000\n","    blk_cords = []\n","    for s,e in blk_chr.iloc[:,2:].values:\n","        blk_cords.extend([c_ for c_ in range(s, e+5000, 5000)])\n","\n","    # blk list : remove intersect of `all_cord`  and `blk`\n","    eligible_coord = [c_ for c_ in all_cord if c_ not in blk_cords]\n","\n","    before_transloc_gene = eligible_coord[eligible_coord <= (spl_s-res)]\n","    after_transloc_gene = eligible_coord[eligible_coord >= spl_s]\n","\n","    if len(before_transloc_gene)>0:\n","        # print('min, max, before:')\n","        # print(min(before_transloc_gene))\n","        # print(max(before_transloc_gene))\n","        before_node_bed = pd.DataFrame({\n","            0: chrid,\n","            1: before_transloc_gene,\n","            2: np.array(before_transloc_gene)+res\n","        })\n","        # save bed files\n","        print(before_node_bed.iloc[:5, ])\n","        before_node_bed.to_csv(os.path.join('results/eligible_coordinates',\n","                                    'eligible_coordinates.before_translocg.{}.hg19.bed'.format(chrid)),\n","                               sep='\\t', header=None, index=None, )\n","\n","    if len(after_transloc_gene)>0:\n","        # print('min, max, after:')\n","        # print(min(after_transloc_gene))\n","        # print(max(after_transloc_gene))\n","        after_node_bed = pd.DataFrame({\n","            0: chrid,\n","            1: after_transloc_gene,\n","            2: np.array(after_transloc_gene)+res\n","        })\n","        # save bed files\n","        print(after_node_bed.iloc[:5, ])\n","        after_node_bed.to_csv(os.path.join('results/eligible_coordinates',\n","                                        'eligible_coordinates.after_translocg.{}.hg19.bed'.format(chrid)), sep='\\t', header=None, index=None, )\n","\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"QnKQiSSbcYcJ","executionInfo":{"status":"ok","timestamp":1678515362917,"user_tz":480,"elapsed":6296,"user":{"displayName":"PY Wu","userId":"03546275994091638450"}},"outputId":"a9f1fe4f-4f81-40bc-cc7c-f1f10df0d6ff"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["chr9\n","20053\n","min, max, before:\n","40000\n","133580000\n","      0      1      2\n","0  chr9  40000  45000\n","1  chr9  45000  50000\n","2  chr9  50000  55000\n","3  chr9  60000  65000\n","4  chr9  85000  90000\n","min, max, after:\n","133590000\n","141110000\n","      0          1          2\n","0  chr9  133590000  133595000\n","1  chr9  133595000  133600000\n","2  chr9  133600000  133605000\n","3  chr9  133605000  133610000\n","4  chr9  133610000  133615000\n","chr22\n","6605\n","min, max, before:\n","16200000\n","23515000\n","       0         1         2\n","0  chr22  16200000  16205000\n","1  chr22  16345000  16350000\n","2  chr22  16360000  16365000\n","3  chr22  16865000  16870000\n","4  chr22  16870000  16875000\n","min, max, after:\n","23525000\n","51230000\n","       0         1         2\n","0  chr22  23525000  23530000\n","1  chr22  23530000  23535000\n","2  chr22  23535000  23540000\n","3  chr22  23540000  23545000\n","4  chr22  23545000  23550000\n"]}]}],"metadata":{"colab":{"provenance":[],"machine_shape":"hm","gpuClass":"premium","authorship_tag":"ABX9TyMsLDUYBTtjhofdVelJ4Qc0"},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"},"accelerator":"GPU","gpuClass":"premium"},"nbformat":4,"nbformat_minor":0}